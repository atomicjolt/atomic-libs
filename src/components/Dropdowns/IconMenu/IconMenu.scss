@use "../../mixins";

.aje-menu {
  position: relative;
  display: inline-block;
  color: var(--text-clr);

  & > button {
    &[aria-expanded="true"] + .aje-menu__dropdown {
      visibility: visible;
      opacity: 1;
      transform: scaleY(1);
    }
  }

  &__dropdown {
    border-radius: 5px;
    box-shadow: 0 1px 3px hsla(0, 0%, 0%, 0.5);
    background-color: var(--neutral50);
    padding: 8px 0;
    z-index: 100;
    transform: scaleY(0.9);
    transform-origin: top;
    transition: transform 200ms ease, opacity 200ms;

    &.drop-right {
      right: initial;
      left: 0;
    }
  }

  &__option {
    position: relative;
    display: flex;
    align-items: center;
    border: none;
    width: 100%;
    gap: 12px;
    padding: 6px 24px 6px 16px;
    min-height: 36px;
    @include mixins.regular;
    line-height: 1;
    color: var(--option-text-clr, inherit);
    font-size: 1.6rem;
    white-space: nowrap;
    background-color: var(--option-bg-clr);

    --option-icon-clr: var(--neutral600);
    i {
      font-size: 2rem;
      color: var(--option-icon-clr);
    }

    &:hover {
      --option-bg-clr: var(--neutral100);
      cursor: pointer;
    }
    &:active {
      --option-bg-clr: var(--neutral200);
    }
    &.is-focused {
      --option-bg-clr: var(--primary700);
      --option-text-clr: var(--text-clr-inverted);
      --option-icon-clr: var(--text-clr-inverted);
      --option-selected-clr: var(--text-clr-inverted);
    }
  }
}
