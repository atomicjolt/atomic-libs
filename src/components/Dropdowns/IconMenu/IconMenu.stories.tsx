import React from "react";
import { ComponentMeta, ComponentStory } from "@storybook/react";
import IconMenuComponent, { IconMenuProps } from ".";

export default {
  title: "Dropdowns/IconMenu",
  component: IconMenuComponent,
} as ComponentMeta<typeof IconMenuComponent>;

const Template: ComponentStory<typeof IconMenuComponent> = (
  args: IconMenuProps
) => (
  <IconMenuComponent {...args}>
    <IconMenuComponent.Item icon="add_alert" onClick={() => {}}>
      Option 1
    </IconMenuComponent.Item>
    <IconMenuComponent.Item icon="info" onClick={() => {}}>
      Option 2
    </IconMenuComponent.Item>
    <IconMenuComponent.Item icon="alarm_on" onClick={() => {}}>
      Option 3
    </IconMenuComponent.Item>
  </IconMenuComponent>
);

export const IconMenu = Template.bind({});
IconMenu.args = {
  icon: "more_vert",
  label: "Custom dropdown label",
  disabled: false,
  position: "bottom-left",
};
