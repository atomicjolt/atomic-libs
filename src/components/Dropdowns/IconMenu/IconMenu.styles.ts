import styled from "styled-components";
import mixins from "../../../styles/mixins";

export const IconMenuWrapper = styled.div`
  position: relative;
  display: inline-block;
  color: var(--text-clr);
`;

export const IconMenuDropdown = styled.div`
  border-radius: 5px;
  box-shadow: 0 1px 3px hsla(0, 0%, 0%, 0.5);
  background-color: var(--neutral50);
  padding: 8px 0;
  z-index: 100;
  transform: scaleY(0.9);
  transform-origin: top;
`;

export const IconMenuOption = styled.button`
  ${mixins.Regular}
  position: relative;
  display: flex;
  align-items: center;
  border: none;
  width: 100%;
  gap: 12px;
  padding: 6px 24px 6px 16px;
  min-height: 36px;
  line-height: 1;
  color: var(--option-text-clr, inherit);
  font-size: 1.6rem;
  white-space: nowrap;
  background-color: var(--option-bg-clr);

  --option-icon-clr: var(--neutral600);
  i {
    font-size: 2rem;
    color: var(--option-icon-clr);
  }

  &:hover {
    --option-bg-clr: var(--neutral100);
    cursor: pointer;
  }
  &:active {
    --option-bg-clr: var(--neutral200);
  }
  &.is-focused {
    --option-bg-clr: var(--primary700);
    --option-text-clr: var(--text-clr-inverted);
    --option-icon-clr: var(--text-clr-inverted);
    --option-selected-clr: var(--text-clr-inverted);
  }
`;
