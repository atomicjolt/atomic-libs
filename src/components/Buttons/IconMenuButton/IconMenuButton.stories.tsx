import React from "react";
import IconMenuButton from ".";
import IconMenu from "../../Dropdowns/IconMenu";
import { Meta, StoryObj } from "@storybook/react";

const meta: Meta<typeof IconMenuButton> = {
  title: "Buttons/IconMenuButton",
  component: IconMenuButton,
  argTypes: {
    onClick: {
      control: false,
      table: {
        category: "Events",
      },
    },
    children: {
      control: "text",
    },
  },
  parameters: {
    layout: "centered",
  },
};

export default meta;

type Story = StoryObj<typeof IconMenuButton>;

export const Default: Story = {
  args: {
    children: "Button",
    type: "button",
    disabled: false,
    loading: false,
    menuIcon: "more_vert",
    menuLabel: "Custom dropdown label",
    menuDisabled: false,
    menuPosition: "bottom-right",
    menuIconVariant: "default",
    menuChildren: <>
      <IconMenu.Item icon="add_alert" onClick={() => { }}>
        Option 1
      </IconMenu.Item>
      <IconMenu.Item icon="info" onClick={() => { }}>
        Option 2
      </IconMenu.Item>
      <IconMenu.Item icon="alarm_on" onClick={() => { }}>
        Option 3
      </IconMenu.Item>
    </>
  }
};

export const Loading: Story = {
  args: {
    ...Default.args,
    loading: true,
    loadingLabel: "Now Loading",
  },
};

export const LoadingComplete: Story = {
  args: {
    ...Default.args,
    loading: true,
    loadingLabel: "Now Loading",
    loadingComplete: true,
  },
};
