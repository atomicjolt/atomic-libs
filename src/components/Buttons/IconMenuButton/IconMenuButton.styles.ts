import styled from "styled-components";
import { BaseStyledButton } from "../Buttons.styles";
import { StyledIconButton } from "../IconButton/IconButton.styles";

export const StyledIconMenuButton = styled.div`
  display: flex;

  & > ${BaseStyledButton} {
    --btn-text-clr: var(--text-clr-alt);
    --btn-bg-clr: var(--neutral50);
    --btn-hover-text-clr: var(--text-clr);
    --btn-hover-bg-clr: var(--neutral100);
    --btn-border: var(--border);

    border-top-right-radius: 0;
    border-bottom-right-radius: 0;
    border-right: 0;
  }

  ${StyledIconButton} {
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }
`;
