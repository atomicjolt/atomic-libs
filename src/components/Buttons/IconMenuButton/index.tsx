import React, { ReactNode } from "react";
import cn from "classnames";
import { ButtonProps } from "../Button";
import Spinner from "../../Loaders/Spinner";
import IconMenu, { IconMenuProps } from "../../Dropdowns/IconMenu";
import { BaseStyledButton } from "../Buttons.styles";
import { StyledIconMenuButton } from "./IconMenuButton.styles";
import { MaterialIconVariants, MaterialIcons } from "../../../types";
import { PopoverPosition } from "../../Utility/Popover/Popover.types";

export type IconMenuButtonProps = Omit<ButtonProps, "variant"> & { 
  menuIcon: MaterialIcons;
  menuLabel: string;
  menuIconVariant?: MaterialIconVariants;
  menuPosition?: PopoverPosition;
  menuDisabled?: boolean;
  menuClassName?: string;
  menuChildren?: ReactNode;
};

export default function IconMenuButton(props: IconMenuButtonProps) {
  const {
    children,
    type = "button",
    disabled = false,
    loading = false,
    loadingLabel = "loading",
    loadingComplete = false,
    onClick,
    className,
    menuIcon,
    menuLabel,
    menuIconVariant,
    menuPosition,
    menuDisabled,
    menuClassName,
    menuChildren,
    ...rest
  } = props;

  const loadingText = loading ? loadingLabel : "";

  return (
    <StyledIconMenuButton>
      <BaseStyledButton
        aria-label={loadingText || undefined}
        type={type}
        className={cn("aje-btn", className, {
          "is-loading": loading,
        })}
        onClick={onClick}
        disabled={disabled}
        {...rest}
      >
        {loading && <Spinner loading={!loadingComplete} />}
        {children}
      </BaseStyledButton>
      <IconMenu label={menuLabel} icon={menuIcon} iconVariant={menuIconVariant} position={menuPosition} disabled={menuDisabled} className={menuClassName} children={menuChildren} />
    </StyledIconMenuButton>
  );
}