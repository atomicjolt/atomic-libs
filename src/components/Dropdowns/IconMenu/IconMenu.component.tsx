import { MenuTriggerProps } from "react-stately";
import { AriaMenuProps } from "@react-aria/menu";
import { Placement } from "@react-aria/overlays";
import { BaseProps, CanHaveIcon, LoadingProps } from "../../../types";
import { Popover } from "../../Overlays/Popover";
import { MenuTrigger } from "../Menu/MenuTrigger";
import { Menu, MenuSectionProps } from "../Menu";
import { IconButton } from "../../Buttons/IconButton";
import { ButtonVariants } from "../../Buttons/Buttons.types";
import { ItemProps } from "@components/Collection";
import { CollectionChildren } from "@react-types/shared";

export type IconMenuProps<T extends object> = AriaMenuProps<T> &
  MenuTriggerProps &
  BaseProps &
  CanHaveIcon &
  LoadingProps & {
    isDisabled?: boolean;
    buttonVariant?: ButtonVariants;
    menuPlacement?: Placement;
    children: CollectionChildren<ItemProps>;
  };

export function IconMenu<T extends {}>(props: IconMenuProps<T>) {
  const {
    icon = "more_vert",
    className,
    iconVariant,
    buttonVariant,
    menuPlacement = "bottom right",
    isDisabled,
    isLoading,
    loadingComplete,
    loadingLabel,
    ...rest
  } = props;

  return (
    <MenuTrigger {...rest}>
      <IconButton
        icon={icon}
        iconVariant={iconVariant}
        variant={buttonVariant}
        className={className}
        isDisabled={isDisabled}
        isLoading={isLoading}
        loadingComplete={loadingComplete}
        loadingLabel={loadingLabel}
      />
      <Popover placement={menuPlacement}>
        <Menu {...rest}>{props.children}</Menu>
      </Popover>
    </MenuTrigger>
  );
}

const IconMenuItem = (props: ItemProps) => <Menu.Item {...props} />;
IconMenuItem.displayName = "IconMenu.Item";
/**
 * @deprecated use `Item` instead
 */
IconMenu.Item = IconMenuItem;

const IconMenuSection = <T extends object>(props: MenuSectionProps<T>) => (
  <Menu.Section {...props} />
);
IconMenuSection.displayName = "IconMenu.Section";
/**
 * @deprecated use `Section` instead
 */
IconMenu.Section = IconMenuSection;

export default IconMenu;
