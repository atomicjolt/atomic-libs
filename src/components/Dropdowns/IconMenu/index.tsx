import React, { useRef } from "react";
import { MenuTriggerProps, useMenuTriggerState } from "react-stately";
import { AriaMenuProps, Placement, useMenuTrigger } from "react-aria";
import IconButton from "../../Buttons/IconButton";
import { BaseProps, HasIcon, LoadingProps } from "../../../types";
import { Popover } from "../../Utility/Popover";
import { Menu } from "../Menu";
import { ButtonVariants } from "../../Buttons/Buttons.types";

export type IconMenuProps<T> = AriaMenuProps<T> &
  MenuTriggerProps &
  BaseProps &
  HasIcon &
  LoadingProps & {
    isDisabled?: boolean;
    buttonVariant?: ButtonVariants;
    placement?: Placement;
  };

export default function IconMenu<T extends {}>(props: IconMenuProps<T>) {
  const state = useMenuTriggerState(props);
  const ref = useRef<HTMLButtonElement>(null);
  const { menuTriggerProps, menuProps } = useMenuTrigger<T>(
    { trigger: props.trigger },
    state,
    ref
  );

  const {
    icon,
    className,
    iconVariant,
    buttonVariant,
    placement = "bottom right",
    isDisabled,
    isLoading,
    loadingComplete,
    loadingLabel,
  } = props;

  return (
    <>
      <IconButton
        icon={icon}
        iconVariant={iconVariant}
        variant={buttonVariant}
        ref={ref}
        className={className}
        isDisabled={isDisabled}
        isLoading={isLoading}
        loadingComplete={loadingComplete}
        loadingLabel={loadingLabel}
        {...menuTriggerProps}
      />
      {state.isOpen && (
        <Popover state={state} triggerRef={ref} placement={placement}>
          <Menu {...props} {...menuProps} />
        </Popover>
      )}
    </>
  );
}
