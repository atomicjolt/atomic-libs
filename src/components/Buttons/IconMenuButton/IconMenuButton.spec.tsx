import React from "react";
import { render } from "@testing-library/react";
import IconMenuButton from ".";
import IconMenu from "../../Dropdowns/IconMenu";

it("matches snapshot", () => {
  const result = render(
    <IconMenuButton iconMenuProps={{
        icon: "more_vert",
        label: "Custom dropdown label",
        disabled: false,
        position: "bottom-left",
        iconVariant: "default",
        children: <>
          <IconMenu.Item icon="add_alert" onClick={() => { }}>
            Option 1
          </IconMenu.Item>
          <IconMenu.Item icon="info" onClick={() => { }}>
            Option 2
          </IconMenu.Item>
          <IconMenu.Item icon="alarm_on" onClick={() => { }}>
            Option 3
          </IconMenu.Item>
        </>
      }}>
        Button
      </IconMenuButton>
  );
  expect(result.asFragment()).toMatchSnapshot();
});
