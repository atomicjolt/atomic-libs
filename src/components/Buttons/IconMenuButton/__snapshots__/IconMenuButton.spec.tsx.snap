// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-iqAbyq juJrWy"
  >
    <button
      class="sc-jSFipO kUfJLg aje-btn"
      type="button"
    >
      Button
    </button>
    <div
      class="sc-pNWRh ikcnyd aje-menu"
    >
      <button
        aria-controls="IconMenu-:r0:-menu"
        aria-expanded="false"
        aria-haspopup="menu"
        aria-label="Custom dropdown label"
        class="sc-jSFipO sc-gKAaef kUfJLg cmuEww aje-btn aje-btn--icon aje-btn--primary"
        id="IconButton-:r1:-button"
        type="button"
      >
        <i
          aria-hidden="true"
          class="sc-eCAqax dRBnmT material-icons is-medium"
        >
          more_vert
        </i>
      </button>
      <div
        class="sc-iCoHzw jzQmut"
      >
        <div
          class="sc-fujznN kNZgZH aje-popover is-auto aje-popover-bottom-right"
        >
          <div
            aria-labelledby="IconMenu-:r0:-button"
            class="sc-jrsKJM eksYmN aje-menu__dropdown"
            id="IconMenu-:r0:-menu"
            role="menu"
            tabindex="-1"
          >
            <button
              class="sc-kEqXeH dsclas aje-menu__option"
            >
              <i
                aria-hidden="true"
                class="sc-eCAqax dRBnmT material-icons is-medium"
              >
                add_alert
              </i>
              Option 1
            </button>
            <button
              class="sc-kEqXeH dsclas aje-menu__option"
            >
              <i
                aria-hidden="true"
                class="sc-eCAqax dRBnmT material-icons is-medium"
              >
                info
              </i>
              Option 2
            </button>
            <button
              class="sc-kEqXeH dsclas aje-menu__option"
            >
              <i
                aria-hidden="true"
                class="sc-eCAqax dRBnmT material-icons is-medium"
              >
                alarm_on
              </i>
              Option 3
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
